plugins/inputs/firehose/firehose.go

```diff
@@ -0,0 +1,413 @@
+	}{
+		RequestID:    r.req.Header.Get(requestIDHeader),
+		Timestamp:    time.Now().Unix(),
+		ErrorMessage: statusCodeToMessage[r.responseStatusCode],
+	}
+	response, err := json.Marshal(responseBody)
+	if err != nil {
+		return err
+	}
+	res.Header().Set("content-type", "application/json")
+	res.WriteHeader(r.responseStatusCode)
+	_, err = res.Write(response)
+	return err
+}
+
+// Firehose is an input plugin that collects external metrics sent via HTTP from AWS Data Firhose
+type Firehose struct {
+	ServiceAddress string          `toml:"service_address"`
+	Paths          []string        `toml:"paths"`
+	PathTag        bool            `toml:"path_tag"`
+	ReadTimeout    config.Duration `toml:"read_timeout"`
+	WriteTimeout   config.Duration `toml:"write_timeout"`
+	AccessKey      string          `toml:"access_key"`
+	ParameterTags  []string        `toml:"parameter_tags"`
+
+	tlsint.ServerConfig
+	tlsConf *tls.Config
+
+	TimeFunc
+	Log telegraf.Logger
+
+	wg    sync.WaitGroup
+	close chan struct{}
+
+	listener net.Listener
+
+	telegraf.Parser
+	acc telegraf.Accumulator
+}
+
+func (*Firehose) SampleConfig() string {
+	return sampleConfig
+}
+
+func (f *Firehose) Gather(_ telegraf.Accumulator) error {
+	return nil
+}
+
+func (f *Firehose) SetParser(parser telegraf.Parser) {
+	f.Parser = parser
+}
+
+// Start starts the http listener service.
+func (f *Firehose) Start(acc telegraf.Accumulator) error {
+	if f.ReadTimeout < config.Duration(time.Second) {
+		f.ReadTimeout = config.Duration(time.Second * 10)
+	}
+	if f.WriteTimeout < config.Duration(time.Second) {
+		f.WriteTimeout = config.Duration(time.Second * 10)
+	}
+
+	f.acc = acc
+
+	server := f.createHTTPServer()
+
+	f.wg.Add(1)
+	go func() {
+		defer f.wg.Done()
+		if err := server.Serve(f.listener); err != nil {
+			if !errors.Is(err, net.ErrClosed) {
+				f.Log.Errorf("Serve failed - %s", err.Error())
+			}
+			close(f.close)
+		}
+	}()
+
+	f.Log.Infof("Listening on %s", f.listener.Addr().String())
+
+	return nil
+}
+
+func (f *Firehose) createHTTPServer() *http.Server {
+	return &http.Server{
+		Addr:         f.ServiceAddress,
+		Handler:      f,
+		ReadTimeout:  time.Duration(f.ReadTimeout),
+		WriteTimeout: time.Duration(f.WriteTimeout),
+		TLSConfig:    f.tlsConf,
+	}
+}
+
+// Stop cleans up all resources
+func (f *Firehose) Stop() {
+	if f.listener != nil {
+		f.listener.Close()
+	}
+	f.wg.Wait()
+}
+
+func (f *Firehose) Init() error {
+	tlsConf, err := f.ServerConfig.TLSConfig()
+	if err != nil {
+		return err
+	}
+
+	var listener net.Listener
+	if tlsConf != nil {
+		listener, err = tls.Listen("tcp", f.ServiceAddress, tlsConf)
+	} else {
+		listener, err = net.Listen("tcp", f.ServiceAddress)
+	}
+	if err != nil {
+		return err
+	}
+	f.tlsConf = tlsConf
+	f.listener = listener
+
+	return nil
+}
+
+func (f *Firehose) ServeHTTP(res http.ResponseWriter, req *http.Request) {
+	select {
+	case <-f.close:
+		res.WriteHeader(http.StatusGone)
+		return
```

Comment:
We do respond to the request using the `res.WriteHeader` method with `http.StatusGone` status code

Reply:
Yes but you could also normally process the request instead of aborting it and exit afterwards (actually that's what should be done when shutting down the server). You do cover a very small time-range here which is a request was made just after we closed the server and just before reaching this code. If you are beyond this statement, the request will be answered anyway...

Reply:
Well, I think then we can just get rid of this. Right?

Reply:
removed